### ISSUE TITLE
Modify Rewrite Middleware RewriteOptions rules collection after Startup runs

### ISSUE THREAD

**Issue by cpkuo** [aspnet/AspNetCore#16658]
Is there a recommended way to modify this rules collection after startup runs without having to store the RewriteOptions as a static variable before passing it to .UseRewriter(options)? The rewrite middleware builder extension method appears to call UseMiddleware<Rewriter>(Options.Create(_options)) but the returning options wrapper is not registered anywhere.

**Comment by Tratcher** [comment]
No, that's not supported. Even storing the RewriteOptions and trying to modify it later is dangerous because that object is being used by multiple threads.

**Comment by xantari** [comment]
I am also trying to do this, because we are building off of a headless CMS, where the rewrite rules are defined in the cloud hosted CMS, but our .NET Application is processing the rewrite rules from that cloud CMS system.

When they make a change in the CMS system, we get a webhook notification of the change, but have no way of updating the rewrite rules.

I tried injecting IOptions<RewriteOptions> into our controller, but the .Values is empty strangely. Even though on app startup we initialize entries into it and they work fine.

This really needs to be supported. Perhaps IOptionsSnapshot will allow you to modify the rewrites on the fly and be thread safe?

**Comment by xantari** [comment]
Also, couldn't you just use a ConcurrentBag or some similar threadsafe collection to store the rules to eliminate the thread safety issue?

**Comment by burtonrodman** [comment]
so we haven't really answered the original question...  what is the recommended way of modifying rewrite rules after startup?  are we just expected to restart the site?

I have also had the thought of using IOptionsSnapshot like @xantari .

Or, if some of the internals (RewriteOptions.BranchedNext and the all the rules) were made public, it would be much easier for anyone to create their own custom middleware.

I am willing to collaborate on a design and submit a PR.
